prow/spyglass/viewers/viewers.go

```diff
@@ -0,0 +1,201 @@
+}
+
+// Artifact represents some output of a prow job
+type Artifact interface {
+	// ReadAt reads len(p) bytes of the artifact at offset off. (unsupported on some compressed files)
+	ReadAt(p []byte, off int64) (n int, err error)
+	// ReadAtMost reads at most n bytes from the beginning of the artifact
+	ReadAtMost(n int64) ([]byte, error)
+	// CanonicalLink gets a link to viewing this artifact in storage
+	CanonicalLink() string
+	// JobPath is the path to the artifact within the job (i.e. without the job prefix)
+	JobPath() string
+	// ReadAll reads all bytes from the artifact up to a limit specified by the artifact
+	ReadAll() ([]byte, error)
+	// ReadTail reads the last n bytes from the artifact (unsupported on some compressed files)
+	ReadTail(n int64) ([]byte, error)
+	// Size gets the size of the artifact in bytes, may make a network call
+	Size() (int64, error)
+	// UseContext allows the consumer to set a context to be used in read operations on this Artifact
+	UseContext(ctx context.Context) error
+}
+
+// ViewHandler consumes artifacts and some possible callback json data and returns an html view.
+// Use the javascript function refreshView(viewName, viewData) to allow your viewer to call back to itself
+// (request more data, update the view, etc.). ViewData is a json blob that will be passed back to the
+// handler function for your view as the string
+type ViewHandler func([]Artifact, string) string
+
+// View gets the updated view from an artifact viewer with the provided name
+func View(name string, artifacts []Artifact, raw string) (string, error) {
+	handler, ok := viewHandlerRegistry[name]
+	if !ok {
+		return "", ErrInvalidViewName
+	}
+	return handler(artifacts, raw), nil
+
+}
+
+// Title gets the title of the view with the given name
+func Title(name string) (string, error) {
+	m, ok := viewMetadataRegistry[name]
+	if !ok {
+		return "", ErrInvalidViewName
+	}
+	return m.Title, nil
+
+}
+
+// Priority gets the priority of the view with the given name
+func Priority(name string) (int, error) {
+	m, ok := viewMetadataRegistry[name]
+	if !ok {
+		return -1, ErrInvalidViewName
+	}
+	return m.Priority, nil
+
+}
+
+// RegisterViewer registers new viewers
+func RegisterViewer(viewerName string, metadata ViewMetadata, handler ViewHandler) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if ok {
+		return fmt.Errorf("viewer already registered with name %s", viewerName)
+	}
+
+	if metadata.Title == "" {
+		return errors.New("empty title field in view metadata")
+	}
+	if metadata.Priority < 0 {
+		return errors.New("priority must be greater than 0")
+	}
+	viewHandlerRegistry[viewerName] = handler
+	viewMetadataRegistry[viewerName] = metadata
+	logrus.Infof("Spyglass registered viewer %s with title %s.", viewerName, metadata.Title)
+	return nil
+}
+
+// UnregisterViewer unregisters viewers
+func UnregisterViewer(viewerName string) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if !ok {
+		return ErrInvalidViewName
```

Comment:
Why do we need to return an error if it isn't registered? `delete(map, missingKey)` will just silently return in this scenario: https://blog.golang.org/go-maps-in-action

Reply:
good point, removed error